Thinking in react translation (#421)
reactjs/it.react.dev: src/content/learn/thinking-in-react.md

]
```

Il mockup appare così:

<img src="/images/docs/s_thinking-in-react_ui.png" width="300" style={{margin: '0 auto'}} />

Per implementare una UI in React, di solito segui i seguenti cinque step.

## Step 1: Dividi la UI in una gerarchia di componenti {/*step-1-break-the-ui-into-a-component-hierarchy*/}

Inizia disegnando rettangoli intorno a ogni componente e sottocomponente nel mockup e assegna loro un nome. Se lavori con un designer, potrebbe avere già nominato questi componenti nel suo tool di design. Chiediglieli!

A seconda del tuo background, puoi pensare a dividere un design in componenti in diversi modi:

> * **Programming**--usa la stesso metodo per decidere se creare una nuova funzione o un nuovo oggetto. Uno di questi metodi è la [single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle), ovvero un componente dovrebbe fare solo una cosa. Se cresce, dovrebbe essere decomposto in sottocomponenti più piccoli.

Author: Assolutamente si! Grazie

Per rendere la UI interattiva, hai bisogno di permettere agli utenti di modificare il tuo modello di dati sottostante. Per questo userai lo *state*.

Immaginati lo state come il set minimo di dati modificabili che la tua app deve ricordare per funzionare. Il principio più importante per la struttura dello state è quello di mantenerlo [DRY (Don't Repeat Yourself).](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) Trova la rappresentazione minima dello state che la tua applicazione ha bisogno e calcola tutto il resto on-demand. Per esempio, se stai costruendo una lista della spesa, puoi memorizzare gli elementi come un array nello state. Se vuoi anche visualizzare il numero di elementi nella lista, non memorizzare il numero di elementi come un altro valore di state--invece, leggi la lunghezza del tuo array.

Adesso immaginati tutti i pezzi di dati in questa applicazione d'esempio:

1. La lista originale di prodotti
2. Il testo di ricerca che l'utente ha inserito
3. Il valore della checkbox
4. La lista filtrata di prodotti

Quale di questi può essere state? Identifica quelli che non lo sono:

* **Rimane invariato** nel tempo? Allora non è state.

> * Viene **passato da un componente genitore** via props? Allora non è state.

Reviewer: ```suggestion
* Viene **passato da un componente genitore** tramite props? Allora non è state.
```

> * **Puoi calcolarlo** basandoti su uno state diverso o da props nel tuo componente? Allora **sicuramente non é** state!

Reviewer: ```suggestion
* **Puoi calcolarlo** basandoti su uno state diverso o da props nel tuo componente? Allora **sicuramente** non è state!
```

> Quello che è rimane probabilmente può considerarsi state.

Reviewer: ```suggestion
Quello che è rimasto probabilmente può considerarsi state.
```

Analizziamo di nuovo questi dati uno alla volta:

> 1. la lista originale di prodotti è **passata in via props, quindi non è state.**

Reviewer: ```suggestion
1. La lista originale di prodotti è **passata tramite props, quindi non è state.**
```

    3. Se non puoi trovare un componente in cui abbia senso possedere lo state, crea un nuovo componente esclusivamente per contenere lo state e aggiungilo in qualche punto della gerarchia sopra il componente genitore comune.

Nello step precedente, hai identificato due pezzi di state in questa applicazione: il testo di input di ricerca e il valore della checkbox. In questo esempio, appaiono sempre insieme, quindi ha senso metterli nello stesso posto.

Adesso eseguiamo la nostra strategia per essi:

1. **Identifica i componenti che usano lo state:**
    * `ProductTable` ha bisogno di filtrare la lista dei prodotti in base a quei valori di state (testo di ricerca e valore della checkbox).
    * `SearchBar` ha bisogno di mostrare lo state (testo di ricerca e valore della checkbox).
1. **Trova il loro genitore comune:** Il primo componente genitore che entrambi i componenti condividono è `FilterableProductTable`.
2. **Decidi dove vive lo state**: Terremo i valori dello state del testo del filtro e della checkbox in `FilterableProductTable`.

Quindi i valori di state vivranno in `FilterableProductTable`.

> Aggiungi lo state al componente con l'[`useState()` Hook.](/reference/react/useState) Gli Hooks sono funzioni speciali che ti permettono di "collegarti" a React. Aggiungi due variabili state all'inizio di `FilterableProductTable` e specifica il loro stato iniziale:

Reviewer: ```suggestion
Aggiungi lo state al componente con l'[Hook `useState()`.](/reference/react/useState) Gli Hooks sono funzioni speciali che ti permettono di "collegarti" a React. Aggiungi due variabili state all'inizio di `FilterableProductTable` e specifica il loro stato iniziale:
```

</div>
```

> Puoi iniziare a vedere come si comporterà la tua applicazione. Modifica il valore iniziale di `filterText` da `useState('')` a `useState('fruit')` nel codice sandbox qui sotto. Vedrai sia il testo di input di ricerca che la tabella si aggiornano:

Reviewer: ```suggestion
Puoi iniziare a vedere come si comporterà la tua applicazione. Modifica il valore iniziale di `filterText` da `useState('')` a `useState('fruit')` nel codice sandbox qui sotto. Vedrai che sia il testo di input di ricerca che la tabella si aggiornano:
```

</Sandpack>

> Nota che la modifica del form non funziona ancora. Nello sandbox sopra c'è un errore di console che spiega perché:

Reviewer: ```suggestion
Nota che la modifica del form non funziona ancora. Nel sandbox sopra c'è un errore di console che spiega perché:
```

<ConsoleBlock level="error">

You provided a \`value\` prop to a form field without an \`onChange\` handler. This will render a read-only field.

</ConsoleBlock>

> Nel sandbox sopra, `ProductTable` e `SearchBar` leggono le props `filterText` e `inStockOnly` per renderizzare la tabella, l'input e checkbox. Ad esempio, ecco come `SearchBar` popola il valore dell'input:

Reviewer: ```suggestion
Nel sandbox sopra, `ProductTable` e `SearchBar` leggono le props `filterText` e `inStockOnly` per renderizzare la tabella, l'input e la checkbox. Ad esempio, ecco come `SearchBar` popola il valore dell'input:
```

        placeholder="Search..."/>
```

Tuttavia, non hai aggiunto alcun codice per rispondere alle azioni dell'utente come la digitazione. Questo sarà il tuo ultimo passaggio.

## Step 5: Aggiungi il flusso di dati inverso {/*step-5-add-inverse-data-flow*/}

Attualmente la tua applicazione viene visualizzata correttamente con le props e lo state che fluiscono verso il basso nella gerarchia. Ma per modificare lo state in base all'input dell'utente, è necessario supportare il flusso di dati anche in senso inverso: i componenti del modulo che si trovano più in basso nell'albero devono aggiornare lo state in `FilterableProductTable`.

> React rende questo flusso di dati in modo esplicito, ma richiede un po' più di scrittura rispetto al two-way data binding. Se provi a digitare o a spuntare la checkbox nell'esempio qui sopra, vedrai che React ignora la tua input. Questo è voluto. Scrivendo `<input value={filterText} />`, hai impostato la prop `value` dell'`input` per essere sempre uguale allo stato `filterText` passato da `FilterableProductTable`. Poiché lo stato `filterText` non viene mai impostato, l'input non cambia mai.

Reviewer: ```suggestion
React rende questo flusso di dati in modo esplicito, ma richiede un po' più di scrittura rispetto al two-way data binding. Se provi a digitare o a spuntare la checkbox nell'esempio qui sopra, vedrai che React ignora il tuo input. Questo è voluto. Scrivendo `<input value={filterText} />`, hai impostato la prop `value` dell'`input` per essere sempre uguale allo stato `filterText` passato da `FilterableProductTable`. Poiché lo stato `filterText` non viene mai impostato, l'input non cambia mai.
```

        onInStockOnlyChange={setInStockOnly} />
```

All'interno di `SearchBar`, aggiungi l'evento `onChange` e imposta lo state del genitore da esso:

> 

Reviewer 2: ```suggestion
```
rimozione extra \n

  onChange={(e) => onFilterTextChange(e.target.value)} />
```

Adesso l'applicazione funziona completamente!

> Now the application fully works!

Reviewer 2: ```suggestion
```
